src/libsync/propagateuploadencrypted.cpp

```diff
@@ -0,0 +1,213 @@
+      job->setProperties({"resourcetype", "http://owncloud.org/ns:fileid"});
+      connect(job, &LsColJob::directoryListingSubfolders, this, &PropagateUploadEncrypted::slotFolderEncryptedIdReceived);
+      connect(job, &LsColJob::finishedWithError, this, &PropagateUploadEncrypted::slotFolderEncryptedIdError);
+      job->start();
+  } else {
+    qCDebug(lcPropagateUpload) << "Folder is not encrypted, getting back to default.";
+    emit folerNotEncrypted();
+  }
+}
+
+
+
+/* We try to lock a folder, if it's locked we try again in one second.
+ * if it's still locked we try again in one second. looping untill one minute.
+ *                                                                      -> fail.
+ * the 'loop':                                                         /
+ *    slotFolderEncryptedIdReceived -> slotTryLock -> lockError -> stillTime? -> slotTryLock
+ *                                        \
+ *                                         -> success.
+ */
+
+void PropagateUploadEncrypted::slotFolderEncryptedIdReceived(const QStringList &list)
+{
+  qCDebug(lcPropagateUpload) << "Received id of folder, trying to lock it so we can prepare the metadata";
+  auto job = qobject_cast<LsColJob *>(sender());
+  const auto& folderInfo = job->_folderInfos.value(list.first());
+  _folderLockFirstTry.start();
+  slotTryLock(folderInfo.fileId);
+}
+
+void PropagateUploadEncrypted::slotTryLock(const QByteArray& fileId)
+{
+  auto *lockJob = new LockEncryptFolderApiJob(_propagator->account(), fileId, this);
+  connect(lockJob, &LockEncryptFolderApiJob::success, this, &PropagateUploadEncrypted::slotFolderLockedSuccessfully);
+  connect(lockJob, &LockEncryptFolderApiJob::error, this, &PropagateUploadEncrypted::slotFolderLockedError);
+  lockJob->start();
+}
+
+void PropagateUploadEncrypted::slotFolderLockedSuccessfully(const QByteArray& fileId, const QByteArray& token)
+{
+  qCDebug(lcPropagateUpload) << "Folder" << fileId << "Locked Successfully for Upload, Fetching Metadata";
+  // Should I use a mutex here?
+  _currentLockingInProgress = true;
+  _folderToken = token;
+  _folderId = fileId;
+
+  auto job = new GetMetadataApiJob(_propagator->account(), _folderId);
+  connect(job, &GetMetadataApiJob::jsonReceived,
+          this, &PropagateUploadEncrypted::slotFolderEncriptedMetadataReceived);
+  job->start();
+}
+
+void PropagateUploadEncrypted::slotFolderEncriptedMetadataReceived(const QJsonDocument &json, int statusCode)
+{
+  qCDebug(lcPropagateUpload) << "Metadata Received, Preparing it for the new file." << json.toVariant();
+
+  // Encrypt File!
+  _metadata = new FolderMetadata(_propagator->account(), json.toJson(QJsonDocument::Compact));
+
+  QFileInfo info(_propagator->_localDir + QDir::separator() + _item->_file);
+  qCDebug(lcPropagateUpload) << "Creating the encrypted file metadata helper.";
+
+  EncryptedFile encryptedFile;
+  encryptedFile.authenticationTag = "NOISE"; // TODO: Remove the noise.
```

Comment:
This needs the tag of the actual file. Else syncing with android fails.

Reply:
So we need to have the encrypted file (in order to have the tag) and then update the metadata. Or update the metadata again afterwards once we have the tag (but that seems silly)